File: dsp08.c
#include <graph.h>
#include <conio.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define XDIM     64
#define ARDIM    64
#define AIDIM    64
#define MAGDIM   64
#define PHASEDIM 64

float   fX[XDIM + 1], fAR[ARDIM + 1], fAI[AIDIM + 1],
        fMAG[MAGDIM + 1], fPHASE[PHASEDIM + 1];

void main (void)
    {
    /*
     *****  PROGRAM NO.8   DISCRETE-TIME FOURIER SERIES ******
     *** 64 REAL SAMPLE VALUES, SPECTRAL MAGNITUDE & PHASE  **
     */
    int    iCount, iCount2;
    float  fMax1, fMax2, fMax3, fB;

    _clearscreen  (_GCLEARSCREEN);

    printf ("PROGRAM 8: Discrete Fourier Series\n");
    printf ("64 real sample values, spectral magnitude ");
    printf ("& phase\n\n");

    printf ("Press any key to start...\n");
    getch();

    for (iCount = 1; iCount <= XDIM; iCount++)
        {
        fB = 2.0 * 3.141593 * (float)(iCount - 1);
        fX[iCount] = sin (fB / 64.0) + cos (fB / 16.0) +
               0.6 * cos (fB / 8.0) + 0.5 * sin (fB / 4.0);
        }

    for (iCount = 1; iCount <= XDIM; iCount++)
        for (iCount2 = 1; iCount2 <= XDIM; iCount2++)
            {
            fAR[iCount] = fAR[iCount] + (1.0 / 64.0) *
                    fX[iCount2] * cos (2.0 * 3.141593 *
                    (float)(iCount - 1) * (float)(iCount2 - 1) / 64.0);
            fAI[iCount] = fAI[iCount] - (1.0 / 64.0) *
                    fX[iCount2] * sin (2.0 * 3.141593 *
                    (float)(iCount - 1) * (float)(iCount2 - 1) / 64.0);
            }

    /*
     ******* CONVERT TO SPECTRAL MAGNITUDE AND PHASE ***********
     */
    for (iCount = 1; iCount <= XDIM; iCount++)
        {
        if (fabs (fAI[iCount]) < 0.001)
            fAI[iCount] = 0.0;
        if (fabs (fAR[iCount]) < 0.00001)
            fAR[iCount] = 0.00001;
        fMAG[iCount] = sqrt (pow (fAR[iCount], 2.0) +
                        pow (fAI[iCount], 2.0));
        fPHASE[iCount] = atan (fAI[iCount] / fAR[iCount]);
        }

    /*
     ********** NORMALISE TO MAXIMUM OF UNITY FOR PLOT *********
     */
    fMax1 = fMax2 = fMax3 = 0.001;

    for (iCount = 1; iCount <= XDIM; iCount++)
        {
        if (fabs (fX[iCount]) > fMax1)
            fMax1 = fabs (fX[iCount]);
        if (fabs (fMAG[iCount]) > fMax2)
             fMax2 = fabs (fMAG[iCount]);
        if (fabs (fPHASE[iCount]) > fMax3)
            fMax3 = fabs (fPHASE[iCount]);
        }

    for (iCount = 1; iCount <= XDIM; iCount++)
        {
        fX[iCount] /= fMax1;
        fMAG[iCount] /= fMax2;
        fPHASE[iCount] /= fMax3;
        }

    /*
     ******** PLOT SIGNAL, SPECTRAL MAGNITUDE AND PHASE ********
     */
    _setvideomode( _MRES16COLOR );

    for (iCount = 1; iCount <= XDIM; iCount++)
      {
      _moveto (5*(iCount - 1), 50);
      _lineto (5*(iCount - 1), (int)(50 - fX[iCount] * 25.0 + 0.5));
      _moveto (5*(iCount - 1), 120);
      _lineto (5*(iCount - 1), (int)(120 - fMAG[iCount] * 30.0 + 0.5));
      _moveto (5*(iCount - 1), 160);
      _lineto (5*(iCount - 1), (int)(160 - fPHASE[iCount] * 20.0 + 0.5));
      }

    getch();
    _setvideomode( _DEFAULTMODE );

    /*
     ***********************************************************
     */
    }

File: dsp08.pas
(********  PROGRAM NO.8   DISCRETE-TIME FOURIER SERIES  ********)
(*****  64 REAL SAMPLE VALUES, SPECTRAL MAGNITUDE & PHASE  *****)
program DSP08;
uses graph3;
var X: array[1..64] of real;
var AR: array[1..64] of real;
var AI: array[1..64] of real;
var MAG: array[1..64] of real;
var PHASE: array[1..64] of real;
var N,K: integer;
var B,M1,M2,M3: real;
begin
  for N:=1 to 64 do
    begin
      AR[N]:=0;AI[N]:=0;
      B:=2*3.141593*(N-1);
      X[N]:=sin(B/64)+cos(B/16)+0.6*cos(B/8)+0.5*sin(B/4);
    end;
  for K:=1 to 64 do
    begin
      for N:=1 to 64 do
        begin
         AR[K]:=AR[K]+(1/64)*X[N]*cos(2*3.141593*(K-1)*(N-1)/64);
         AI[K]:=AI[K]-(1/64)*X[N]*sin(2*3.141593*(K-1)*(N-1)/64);
        end;
    end;
(***********  CONVERT TO SPECTRAL MAGNITUDE & PHASE  ***********)
  for K:=1 to 64 do
    begin
      if abs(AI[K])<0.001 then AI[K]:=0;
      if abs(AR[K])<0.00001 then AR[K]:=0.00001;
      MAG[K]:=sqrt(AR[K]*AR[K]+AI[K]*AI[K]);
      PHASE[K]:=arctan(AI[K]/AR[K]);
    end;
(*******  NORMALISE TO MAXIMUM VALUE OF UNITY FOR PLOT  ********)
  M1:=0.001;M2:=0.001;M3:=0.001;
    for N:=1 to 64 do
      begin
        if abs(X[N])>M1 then M1:=abs(X[N]);
        if abs(MAG[N])>M2 then M2:=abs(MAG[N]);
        if abs(PHASE[N])>M3 then M3:=abs(PHASE[N]);
      end;
  for N:=1 to 64 do
    begin
      X[N]:=X[N]/M1;MAG[N]:=MAG[N]/M2;
      PHASE[N]:=PHASE[N]/M3;
    end;
(********  PLOT SIGNAL, SPECTRAL MAGNITUDE, AND PHASE  *********)
  graphmode;
  for N:=1 to 64 do
    begin
      draw(5*(N-1),50,5*(N-1),round(50-X[N]*25),1);
      draw(5*(N-1),120,5*(N-1),round(120-MAG[N]*30),1);
      draw(5*(N-1),160,5*(N-1),round(160-PHASE[N]*20),1);
    end;
end.
(***************************************************************)